pagerduty-xbar/src/main.rs

```diff
@@ -53,36 +39,77 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
         println!("{icon_not_on_duty}");
     }
 
+    println!("---");
+    println!("Upcoming:");
+
+    for oncall in &oncalls {
+        let delta = oncall.start - now;
+
+        let time_remaining = if delta <= TimeDelta::zero() {
+            "now".into()
+        } else {
+            let days = delta.num_days();
+            if days > 0 {
+                format!("{}d", days)
+            } else {
+                "<1d".into()
+            }
+        };
+
+        println!(
+            "{} - ({}) {} - {}",
+            oncall.schedule.summary,
+            time_remaining,
+            format_readable(oncall.start),
+            format_readable(oncall.end)
+        );
+    }
+
     Ok(())
 }
 
-/// https://developer.pagerduty.com/api-reference/e1ad560792567-list-users-on-call
-async fn check_if_user_on_call(
-    shift_from: DateTime<Utc>,
-    shift_to: DateTime<Utc>,
+fn is_oncall(time: DateTime<Utc>, oncalls: &[Oncall]) -> bool {
+    oncalls.iter().any(|o| time >= o.start && time < o.end)
+}
+
+fn format_readable(dt: DateTime<Utc>) -> String {
+    let local: DateTime<Local> = dt.into();
+    local.format("%a %d-%b-%Y %H:%M").to_string()
+}
+
+fn format_iso(dt: DateTime<Utc>) -> String {
+    dt.to_rfc3339_opts(SecondsFormat::Secs, true)
+}
+
+/// https://developer.pagerduty.com/api-reference/3a6b910f11050-list-all-of-the-on-calls
+async fn get_oncalls(
+    from: DateTime<Utc>,
+    to: DateTime<Utc>,
     pg_auth_token: &str,
     pg_schedule_id: &str,
-    pg_user: &str,
-) -> Result<bool, Box<dyn std::error::Error>> {
-    let since_iso = shift_from.to_rfc3339_opts(SecondsFormat::Secs, true);
-    let until_iso = shift_to.to_rfc3339_opts(SecondsFormat::Secs, true);
+    pg_user_id: &str,
+) -> Result<Vec<Oncall>, Box<dyn std::error::Error>> {
     let client = Client::new();
-    let mut headers = HeaderMap::new();
-    headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
-    let authorization = format!("Token token={pg_auth_token}");
-    headers.insert(AUTHORIZATION, HeaderValue::from_str(&authorization)?);
-    let url = format!(
-        "https://api.pagerduty.com/schedules/{pg_schedule_id}/users?since={since_iso}&until={until_iso}"
-    );
-    eprintln!("{url}");
-    let res = client.get(url).headers(headers).send().await?;
-
-    let response_body = res.text().await?;
-    eprintln!("{response_body}");
```

Comment:
I removed these because the output was very noisy, I'm fine with adding them back in. FYI `dbg!()` is usually more convenient than just printing strings.